keps/sig-cli/20190318-kustomize-generators-transformers.md

```diff
@@ -0,0 +1,395 @@
+   those solutions with kustomize and kubectl.
+1. Enable tools and authoring solutions developed by the ecosystem to natively integrate with kustomize and kubectl
+   commands.
+1. Enable Kustomize power users to augment Kustomize with their own *Transformers* and *Generators*.
+
+### Background
+
+Many users of Kubernetes require the ability to also write Config through high-level abstractions
+rather than only as low-level APIs.
+
+- Numerous tools and DSLs have been developed in the ecosystem in order to address this need
+  - Helm
+  - Ksonnet
+- Organizations have built their own tooling for generating Kubernetes Resource Config
+  - [kube-gen] (AirBnB) 
+
+Support for connecting the tools developed by the ecosystem with kubectl relies on piping commands together,
+however pipes are an imperative technique.  This KEP proposes a declarative approach for composing tools built
+in the Kubernetes ecosystem or bespoke tools developed by organizations that are tightly coupled to their
+environments.
+
+[kube-gen]: https://events.linuxfoundation.org/wp-content/uploads/2017/12/Services-for-All-How-To-Empower-Engineers-with-Kubernetes-Melanie-Cebula-Airbnb.pdf
+
+### Goals
+
+- Allow config authoring solutions developed in the ecosystem to be integrated into kubectl as declarative
+  Generators and Transformers
+- Allow users and organizations to develop their own custom config authoring solutions and integrate
+  them into kubectl
+- Allow Kustomize power users to augment Kustomize's own Tranformers with their own
+- Allow users to perform client-side mutations so they they show up in diffs and are auditable
+
+### Non-Goals
+
+- Rewrite Kustomize on top of plugins
+  - This should be possible, but isn't a problem we need to solve right now
+- Develop a market place of any sort
+  - The ecosystem solutions themselves will have market places.  This allows those ecosystem tools to
+    plug their market places into kubectl via Kustomize.
+
+## Proposal
+
+Introduce 2 new types of `exec` plugins:
+
+- Transformers
+- Generators
+
+#### Terms
+
+- *Virtual Resource*
+  - Resource that is not a server-side Kubernetes API object, but used to configure tools
+  - Examples: kubeconfig, kustomization
+- *Generator*
+  - Kustomize directive that accepts a Virtual Resource as input, and emits non-Virtual Resources as output
+  - Examples: `configMapGenerator`, `secretGenerator`
+- *Transformer*
+  - Kustomize directive that accepts 1 Virtual Resource *and* collection of non-Virtual Resources as input
+    and emits non-Virtual Resources as output.
+  - Examples: `commonAnnotations`, `namespace`
+- *Built-In*
+  - Either a Transformer or Generator that is embedded directly in the `kustomization.yaml` rather than
+    as a separate API.
+
+#### Generators
+
+Generators generate 0 or more Resources for some virtual Resource input.
+
+Generator plugins are run immediately after built-in generators:
```

Comment:
Why not just run them in the order that they are listed?

Reply:
there shouldn't be an order to fields.